crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -0,0 +1,262 @@
+                        };
+                        apply_factor(executed_in_surplus_token, factor)
+                            .ok_or(Error::Factor(executed_in_surplus_token, factor))?
+                    },
+                )),
+                order::FeePolicy::PriceImprovement {
+                    factor: _,
+                    max_volume_factor: _,
+                    quote: _,
+                } => Err(Error::UnimplementedFeePolicy),
+                order::FeePolicy::Volume { factor: _ } => Err(Error::UnimplementedFeePolicy),
+            }
+        };
+
+        let protocol_fee = self.policies.first().map(protocol_fee).transpose();
+        Ok(eth::Asset {
+            token: self.surplus_token(),
+            amount: protocol_fee?.unwrap_or(0.into()).into(),
+        })
+    }
+
+    /// Protocol fee is defined by fee policies attached to the order.
+    ///
+    /// Denominated in NATIVE token
+    fn native_protocol_fee(&self, prices: &auction::Prices) -> Result<eth::TokenAmount, Error> {
+        let protocol_fee = self
+            .surplus_token_price(prices)?
+            .apply(self.protocol_fee()?.amount);
+        // normalize
+        Ok((protocol_fee.0 / *UNIT).into())
+    }
+
+    fn surplus_token(&self) -> eth::TokenAddress {
+        match self.side {
+            Side::Buy => self.sell.token,
+            Side::Sell => self.buy.token,
+        }
+    }
+
+    /// Returns the price of the trade surplus token
+    fn surplus_token_price(&self, prices: &auction::Prices) -> Result<auction::Price, Error> {
+        prices
+            .get(&self.surplus_token())
+            .cloned()
+            .ok_or(Error::MissingPrice(self.surplus_token()))
+    }
+}
+
+fn apply_factor(amount: eth::U256, factor: f64) -> Option<eth::U256> {
+    Some(
+        amount.checked_mul(eth::U256::from_f64_lossy(factor * 1000000000000000000.))?
+            / 1000000000000000000u128,
+    )
+}
+
+/// Custom clearing prices at which the trade was executed.
+///
+/// These prices differ from uniform clearing prices, in that they are adjusted
+/// to account for all fees (gas cost and protocol fees).
+///
+/// These prices determine the actual traded amounts from the user perspective.
+#[derive(Debug, Clone)]
+pub struct CustomClearingPrices {
+    pub sell: eth::U256,
+    pub buy: eth::U256,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("multiple fee policies are not supported yet")]
+    MultipleFeePolicies,
+    #[error("fee policy not implemented yet")]
+    UnimplementedFeePolicy,
+    #[error("failed to calculate surplus for trade sell {0:?} buy {1:?}")]
+    Surplus(eth::Asset, eth::Asset),
+    #[error("missing native price for token {0:?}")]
+    MissingPrice(eth::TokenAddress),
+    #[error("factor {1} multiplication with {0} failed")]
+    Factor(eth::U256, f64),
+    #[error(transparent)]
+    Math(#[from] super::Math),
```

Comment:
Can we replace Factor and use Math in those cases (Factor can only happen in case of Overflow)?

Conceptually, this Math error is equivalent to `solution::Error::Math`, so I think it would even make sense to refactor our solution domain's error model. However I don't think this is blocking this PR.